File: rust/rust-coding-challenges/src/leetcode/find_duplicate_subtrees.rs
use std::{cell::RefCell, collections::HashMap, rc::Rc};

/**
 * Given the root of a binary tree, return all duplicate subtrees.
 * For each kind of duplicate subtrees, you only need to return the root node of any one of them.
 * Two trees are duplicate if they have the same structure with the same node values.
 *
 * Example 1:
 *              1
 *            /   \
 *           2     3
 *         /      /   \
 *        4      2      4
 *              /
 *             4
 *
 *  Input: root = [1,2,3,4,null,2,4,null,null,4]
 *  Output: [[2,4],[4]]
 *
 *
 * Example 2:
 *              2
 *            /    \
 *          1        1
 *
 *  Input: root = [2,1,1]
 *  Output: [[1]]
 *
 *
 * Example 3:
 *              2
 *            /   \
 *          2       2
 *        /       /
 *       3       3
 *
 *  Input: root = [2,2,2,3,null,3,null]
 *  Output: [[2,3],[3]]
 *
 *
 * Constraints:
 *      The number of the nodes in the tree will be in the range [1, 5000]
 *      -200 <= Node.val <= 200
 */
#[derive(Debug, PartialEq, Eq)]
struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}
type TreeNodeArr = Vec<Option<Rc<RefCell<TreeNode>>>>;
pub type TreeNodeOption = Option<Rc<RefCell<TreeNode>>>;

mod BruteForceSolution {
    use super::{TreeNodeArr, TreeNodeOption};
    use crate::leetcode::find_duplicate_subtrees::HashMap;

    /** Brute force solution has time complexity of O(n) */
    fn find_duplicate_subtrees_brute_force(root: TreeNodeOption) -> TreeNodeArr {
        let mut result: TreeNodeArr = Vec::new();
        post_order_brute_force(&root, &mut HashMap::new(), &mut result);
        return result;
    }

    fn post_order_brute_force(
        root: &TreeNodeOption,
        map: &mut HashMap<String, i32>,
        result: &mut TreeNodeArr,
    ) -> String {
        match root {
            None => return String::from("#"),
            Some(currentNode) => {
              
                let currentNode = currentNode.as_ref().borrow();

                let serial = format!(
                    "{},{},{}",
                    currentNode.val,
                    post_order_brute_force(&currentNode.left, map, result),
                    post_order_brute_force(&currentNode.right, map, result)
                );

                let serialCount = map.entry(serial.to_owned()).or_insert(0);
                *serialCount += 1;

                if Some(&(2 as i32)) == map.get(&serial) {
                    result.push(root.to_owned());
                }
                return serial;
            }
        }
    }

    pub fn run_demo(
        input_tree_1: TreeNodeOption,
        input_tree_2: TreeNodeOption,
        input_tree_3: TreeNodeOption,
    ) {
        println!("Brute force Solution");
        println!(
            "input tree 1 solution: \n{:?}",
            find_duplicate_subtrees_brute_force(input_tree_1)
        );

        println!(
            "input tree 2 solution: \n{:?}",
            find_duplicate_subtrees_brute_force(input_tree_2)
        );

        println!(
            "input tree 3 solution: \n{:?}",
            find_duplicate_subtrees_brute_force(input_tree_3)
        );
    }
}

mod UsingSerialIdSolution {
    use std::collections::HashMap;

    use super::{TreeNodeArr, TreeNodeOption};

    static  mut CUR_ID: i32 = 1;

    fn find_duplicate_subtrees_using_serial_ids(root: TreeNodeOption) -> TreeNodeArr {
        let mut serialToIdMap: HashMap<String, i32> = HashMap::new();
        let mut idToCountMap: HashMap<i32, i32> = HashMap::new();
        let mut res: TreeNodeArr = Vec::new();
        post_order_using_serial_ids(&root, &mut serialToIdMap, &mut idToCountMap, &mut res);
        return res;
    }

    fn post_order_using_serial_ids(
        root: &TreeNodeOption,
        serialToIdMap: &mut HashMap<String, i32>,
        idToCountMap: &mut HashMap<i32, i32>,
        result: &mut TreeNodeArr,
    ) -> i32 {
        match root {
            None => return 0,
            Some(currentNode) => {
                let currentNode = currentNode.as_ref().borrow();
                let leftId =
                    post_order_using_serial_ids(&currentNode.left, serialToIdMap, idToCountMap, result, );
                let rightId =
                    post_order_using_serial_ids(&currentNode.right, serialToIdMap, idToCountMap, result,);

                let cur_serial = format!("{},{}.{}", leftId, currentNode.val, rightId);

                unsafe {
                  let serial_id = serialToIdMap.get(&cur_serial).get_or_insert(&CUR_ID).to_owned();

                if serial_id == CUR_ID {
                    CUR_ID = CUR_ID + (1 as i32);
                };

                serialToIdMap.insert(cur_serial.to_owned(), serial_id.to_owned());

                let id_count = idToCountMap.entry(serial_id.to_owned()).or_insert(0);
                *id_count += 1;

                if idToCountMap.get(&serial_id) == Some(&(2 as i32)) {
                    result.push(root.to_owned());
                };
                return serial_id.to_owned();
                }
                

            }
        }
    }

    pub fn run_demo(
        input_tree_1: TreeNodeOption,
        input_tree_2: TreeNodeOption,
        input_tree_3: TreeNodeOption,
    ) {
        println!("Serial ID Solution");
        println!(
            "input tree 1 solution: \n{:?}",
            find_duplicate_subtrees_using_serial_ids(input_tree_1)
        );

        println!(
            "input tree 2 solution: \n{:?}",
            find_duplicate_subtrees_using_serial_ids(input_tree_2)
        );

        println!(
            "input tree 3 solution: \n{:?}",
            find_duplicate_subtrees_using_serial_ids(input_tree_3)
        );
    }
}

pub fn run_demo() {
    let input_tree_1: TreeNodeOption = Some(Rc::new(RefCell::new(TreeNode {
        val: 1,
        left: Some(Rc::new(RefCell::new(TreeNode {
            val: 2,
            left: Some(Rc::new(RefCell::new(TreeNode::new(4)))),
            right: None,
        }))),
        right: Some(Rc::new(RefCell::new(TreeNode {
            val: 3,
            left: Some(Rc::new(RefCell::new(TreeNode {
                val: 2,
                left: Some(Rc::new(RefCell::new(TreeNode::new(4)))),
                right: None,
            }))),
            right: Some(Rc::new(RefCell::new(TreeNode::new(4)))),
        }))),
    })));

    let input_tree_2: TreeNodeOption = Some(Rc::new(RefCell::new(TreeNode {
        val: 2,
        left: Some(Rc::new(RefCell::new(TreeNode::new(1)))),
        right: Some(Rc::new(RefCell::new(TreeNode::new(1)))),
    })));

    let input_tree_3: TreeNodeOption = Some(Rc::new(RefCell::new(TreeNode {
        val: 2,
        left: Some(Rc::new(RefCell::new(TreeNode {
            val: 2,
            left: Some(Rc::new(RefCell::new(TreeNode::new(3)))),
            right: None,
        }))),
        right: Some(Rc::new(RefCell::new(TreeNode {
            val: 2,
            left: Some(Rc::new(RefCell::new(TreeNode::new(3)))),
            right: None,
        }))),
    })));

    BruteForceSolution::run_demo(input_tree_1, input_tree_2, input_tree_3)
}

struct  Solution{}

impl Solution {
    pub fn find_duplicate_subtrees(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Option<Rc<RefCell<TreeNode>>>> {
             let mut result: TreeNodeArr = Vec::new();
        Solution::post_order(&root, &mut HashMap::new(), &mut result);
        return result;
    }

    pub fn post_order(
        root: &TreeNodeOption,
        map: &mut HashMap<String, i32>,
        result: &mut TreeNodeArr,
    ) -> String {
        match root {
            None => return String::from("#"),
            Some(currentNode) => {
              
                let currentNode = currentNode.as_ref().borrow();

                let serial = format!(
                    "{},{},{}",
                    currentNode.val,
                    Solution::post_order(&currentNode.left, map, result),
                    Solution::post_order(&currentNode.right, map, result)
                );

                let serialCount = map.entry(serial.to_owned()).or_insert(0);
                *serialCount += 1;

                if Some(&(2 as i32)) == map.get(&serial) {
                    result.push(root.to_owned());
                }
                return serial;
            }
        }
    }

}

File: typescript/leet-code/find-duplicate-subtrees.ts
/**
 * Given the root of a binary tree, return all duplicate subtrees.
 * For each kind of duplicate subtrees, you only need to return the root node of any one of them.
 * Two trees are duplicate if they have the same structure with the same node values.
 *
 * Example 1:
 *              1
 *            /   \
 *           2     3
 *         /      /   \
 *        4      2      4
 *              /
 *             4
 *
 *  Input: root = [1,2,3,4,null,2,4,null,null,4]
 *  Output: [[2,4],[4]]
 *
 *
 * Example 2:
 *              2
 *            /    \
 *          1        1
 *
 *  Input: root = [2,1,1]
 *  Output: [[1]]
 *
 *
 * Example 3:
 *              2
 *            /   \
 *          2       2
 *        /       /
 *       3       3
 *
 *  Input: root = [2,2,2,3,null,3,null]
 *  Output: [[2,3],[3]]
 *
 *
 * Constraints:
 *      The number of the nodes in the tree will be in the range [1, 5000]
 *      -200 <= Node.val <= 200
 */
/**
 * Definition for a binary tree node.
 * */
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

type TreeNodeArr = Array<TreeNode | null>;

/** Brute force solution has time complexity of O(n) */
function findDuplicateSubtreesBruteForce(root: TreeNode | null): TreeNodeArr {
  let result: TreeNodeArr = [];
  postOrderBruteForce(root, new Map<string, number>(), result);
  return result;
}

function postOrderBruteForce(
  currentNode: TreeNode | null,
  map: Map<string, number>,
  result: TreeNodeArr
): string {
  if (currentNode == null) return "#";
  let serial: string =
    currentNode.val +
    "," +
    postOrderBruteForce(currentNode.left, map, result) +
    "," +
    postOrderBruteForce(currentNode.right, map, result);
  let serialCount = map.get(serial);

  if (serialCount) {
    map.set(serial, ++serialCount);
  } else {
    map.set(serial, 1);
  }

  if (map.get(serial) == 2) result.push(currentNode);
  return serial;
}

let curId = 1;

function findDuplicateSubtreesUsingSerialIds(
  root: TreeNode | null
): TreeNodeArr {
  let serialToIdMap = new Map<string, number>();
  let idToCountMap = new Map<number, number>();
  let res: TreeNodeArr = [];
  postOrderUsingSerialIds(root, serialToIdMap, idToCountMap, res);
  return res;
}

function postOrderUsingSerialIds(
  root: TreeNode | null,
  serialToIdMap: Map<string, number>,
  idToCountMap: Map<number, number>,
  result: TreeNodeArr
): number {
  if (root == null) return 0;
  let leftId = postOrderUsingSerialIds(
    root.left,
    serialToIdMap,
    idToCountMap,
    result
  );
  let rightId = postOrderUsingSerialIds(
    root.right,
    serialToIdMap,
    idToCountMap,
    result
  );

  let curSerial = leftId + "," + root.val + "," + rightId;
  let serialId = serialToIdMap.get(curSerial) ?? curId;

  if (serialId == curId) curId++;

  serialToIdMap.set(curSerial, serialId);

  idToCountMap.set(serialId, (idToCountMap.get(serialId) ?? 0) + 1);

  if (idToCountMap.get(serialId) === 2) result.push(root);
  return serialId;
}

let inputTree1 = new TreeNode(
  1,
  new TreeNode(2, new TreeNode(4)),
  new TreeNode(3, new TreeNode(2, new TreeNode(4)), new TreeNode(4))
);
let inputTree2 = new TreeNode(2, new TreeNode(1), new TreeNode(1));
let inputTree3 = new TreeNode(
  2,
  new TreeNode(2, new TreeNode(3)),
  new TreeNode(2, new TreeNode(3))
);
console.log("Brute force Solution");
console.log(
  "input tree 1 solution: \n",
  findDuplicateSubtreesBruteForce(inputTree1)
);

console.log(
  "input tree 2 solution: \n",
  findDuplicateSubtreesBruteForce(inputTree2)
);

console.log(
  "input tree 3 solution: \n",
  findDuplicateSubtreesBruteForce(inputTree3)
);

console.log("Serial ID Solution");
console.log(
  "input tree 1 solution: \n",
  findDuplicateSubtreesUsingSerialIds(inputTree1)
);

console.log(
  "input tree 2 solution: \n",
  findDuplicateSubtreesUsingSerialIds(inputTree2)
);

console.log(
  "input tree 3 solution: \n",
  findDuplicateSubtreesUsingSerialIds(inputTree3)
);

export {}